### PR TITLE
Add support to include c2p authority

### PR DESCRIPTION
**Pull request by arvindbr8**
As part of xds federation changes, we need to add support to the bootstrap generator to add c2p authority to the config. This change addresses that. 

### what's happenin?
1. add C2P TD authority option using a `include-directpath-authority-experimental` flag 
2. add additional logic to the bootstrap generator to query `http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s` and set the `TRAFFICDIRECTOR_DIRECTPATH_C2P_IPV6_CAPABLE` to true or false, in node metadata

Part of #32

### CODE REVIEW

**Comment by arvindbr8** [comment]
@easwars please take a look

File: main.go [hunk 1]
```diff
@@ -388,6 +395,10 @@ type server struct {
 	ServerFeatures []string `json:"server_features,omitempty"`
 }
 
+// Authority
```
Comment (easwars) [inline main.go:398]: Please add a complete comment.

File: main.go [hunk 2]
```diff
@@ -268,10 +272,13 @@ func generate(in configInput) ([]byte, error) {
 		// Authorities with an empty server config will end up using
 		// the top-level server config. For more details, see:
 		// https://github.com/grpc/proposal/blob/master/A47-xds-federation.md#bootstrap-config-changes.
-		c.Authorities = map[string]struct{}{
+		c.Authorities = map[string]Authority{
 			tdURI: {},
 			"":    {},
 		}
+		if in.includeC2PAuthority {
+			c.Authorities[c2pAuthority] = Authority{}
```
Comment (easwars) [inline main.go:280]: We don't want this to be empty, right. If it is empty, this authority will also use the top-level server config which will send the xDS requests to `trafficdirector.googleapis.com:443`. We want requests matching this authority to go to `dns:///directpath-pa.googleapis.com` instead.

**Review by easwars — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (arvindbr8) [reply main.go:398]: it didnt let me do the TODO without the `authority` in there. i'll add something in

File: main.go (hunk 2 above)
Comment (arvindbr8) [reply main.go:280]: oh right and use DP. I think I was just talking to @apolycn about it now

File: main.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ import (
 )
 
 const tdURI = "trafficdirector.googleapis.com:443"
+const c2pAuthority = "traffic-director-c2p.xds.googleapis.com"
```
Comment (easwars) [inline main.go:36]: Nit: create a const block?

```golang
const (
    tdURI = "trafficdirector.googleapis.com:443"
    c2pAuthority = "traffic-director-c2p.xds.googleapis.com"
)
```

File: main.go [hunk 4]
```diff
@@ -51,6 +52,7 @@ var (
 	gceVM                    = flag.String("gce-vm-experimental", "", "GCE VM name to use, instead of reading it from the metadata server. This flag is EXPERIMENTAL and may be changed or removed in a later release.")
 	configMesh               = flag.String("config-mesh-experimental", "", "Dictates which Mesh resource to use. This flag is EXPERIMENTAL and may be changed or removed in a later release.")
 	includeFederationSupport = flag.Bool("include-federation-support-experimental", false, "whether or not to generate configs required for xDS Federation. This flag is EXPERIMENTAL and may be changed or removed in a later release.")
+	includeC2PAuthority      = flag.Bool("include-c2p-authority-experimental", false, "whether or not to include c2p TD authority for xDS Federation. Ignored if not used with include-federation-support-experimental flag. This flag is EXPERIMENTAL and may be changed or removed in a later release.")
```
Comment (easwars) [inline main.go:55]: `c2p` might not be very descriptive for the users. Maybe `include-directpath-authority-experimental`. Wdyt @ejona86 ?

File: main.go [hunk 5]
```diff
@@ -388,6 +404,12 @@ type server struct {
 	ServerFeatures []string `json:"server_features,omitempty"`
 }
 
+// Authority is the corresponding configuration to an authority name in the map.
```
Comment (easwars) [inline main.go:407]: Maybe:

```golang
// Authority is the configuration corresponding to an authority name in the map.
//
// For more details, see:
// https://github.com/grpc/proposal/blob/master/A47-xds-federation.md#bootstrap-config-changes
```

File: main.go [hunk 6]
```diff
@@ -268,10 +272,22 @@ func generate(in configInput) ([]byte, error) {
...
 			tdURI: {},
 			"":    {},
 		}
+		if in.includeC2PAuthority {
+			c.Authorities[c2pAuthority] = Authority{
+				XdsServers: []server{
+					{
+						ServerUri: "dns:///directpath-pa.googleapis.com",
+						ChannelCreds: []creds{
+							{Type: "google_default"},
```
Comment (easwars) [inline main.go:285]: We need server features with `v3` support for c2p TD as well.

File: main.go (hunk 3 above)
Comment (arvindbr8) [reply main.go:36]: done :)

File: main.go (hunk 5 above)
Comment (arvindbr8) [reply main.go:407]: yup, that makes sense

File: main.go (hunk 6 above)
Comment (arvindbr8) [reply main.go:285]: sure, updated. So do we also want to check for `ignore_resource_deletion` too?

**Comment by arvindbr8** [comment]
I'm also updating this to include changes to support adding node metadata indicating IPv6 capabilities for DirectPath [b/258291928](https://b.corp.google.com/issues/258291928)

File: main.go (hunk 6 above)
Comment (easwars) [reply main.go:285]: Hmm ... maybe it might be a good idea to pull this out into a helper function. That way, we can be sure that we use the same logic to generate the server config for both TDs.

File: main.go [hunk 7]
```diff
@@ -203,12 +211,7 @@ func validate(in configInput) error {
 func generate(in configInput) ([]byte, error) {
 	c := &config{
 		XdsServers: []server{
-			{
-				ServerUri: in.xdsServerUri,
-				ChannelCreds: []creds{
-					{Type: "google_default"},
-				},
-			},
+			generateServerConfigFromInput(in.xdsServerUri, in),
```
Comment (easwars) [inline main.go:214]: I'm wondering if we could make `generateServerConfigFromInput` return a `[]server` with a single entry instead of the caller doing that part. If we ever have to end up supporting more than one entry in the server config, we can deal with that then. What do you think?

File: main.go [hunk 8]
```diff
@@ -347,6 +355,15 @@ func getVMName() string {
 	return string(vm)
 }
 
+// isIPv6Capable. Retrieve IPv6 address (if any) from metadata server to check
+// for DirectPath IPv6 capability.
+func isIPv6Capable() bool {
+	if _, err := getFromMetadata("http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"); err == nil {
+		return true
+	}
+	return false
```
Comment (easwars) [inline main.go:364]: Optional:

```golang
    _, err := getFromMetadata("http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"); 
    return err == nil
```

File: main_test.go [hunk 9]
```diff
@@ -544,6 +605,24 @@ func TestGetVMName(t *testing.T) {
 	}
 }
 
+func TestCheckIPv6Capable(t *testing.T) {
+	server := httptest.NewServer(nil)
+	defer server.Close()
+	overrideHTTP(server)
+	want := true
```
Comment (easwars) [inline main_test.go:612]: Nit: move `want` to be declared right above where it is used.

File: main_test.go [hunk 10]
```diff
@@ -544,6 +605,24 @@ func TestGetVMName(t *testing.T) {
...
+	http.HandleFunc("http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s",
+		func(w http.ResponseWriter, r *http.Request) {
+			if r.Header.Get("Metadata-Flavor") != "Google" {
+				http.Error(w, "Missing Metadata-Flavor", 403)
+				return
+			}
+			w.Write([]byte("6970:7636:2061:6464:7265:7373:2062:6162"))
+		})
+	if got := isIPv6Capable(); got != want {
```
Comment (easwars) [inline main_test.go:621]: Should we have a test case where `isIPv6Capable()` returns false?

I see that other tests don't do that. I don't remember why. But, if we *can* test both cases here, that would be good I think.

**Review by easwars — COMMENTED** [review]
I think changes are mostly looking good. Some very minor comments this time around.

File: main.go (hunk 8 above)
Comment (arvindbr8) [reply main.go:364]: looks better.. im updating this

File: main_test.go (hunk 9 above)
Comment (arvindbr8) [reply main_test.go:612]: makes sense.. updating this

File: main.go [hunk 11]
```diff
@@ -388,6 +399,36 @@ type server struct {
...
+	return ss
+}
+
+// Authority is the configuration corresponding to an authority name in the map.
+//
+// For more details, see:
+// https://github.com/grpc/proposal/blob/master/A47-xds-federation.md#bootstrap-config-changes
+type Authority struct {
+	XdsServers                         []server `json:"xds_servers,omitempty"`
+	ClientListenerResourceNameTemplate string   `json:"client_listener_resource_name_template,omitempty"`
```
Comment (easwars) [inline main.go:429]: Do we need this field at this point, given that we are not using it?

File: main_test.go [hunk 12]
```diff
@@ -544,6 +605,49 @@ func TestGetVMName(t *testing.T) {
...
+			},
+			wantOutput: false,
+		},
+	}
+	server := httptest.NewServer(nil)
+	defer server.Close()
+	overrideHTTP(server)
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			uuid.SetRand(rand.New(rand.NewSource(1)))
```
Comment (easwars) [inline main_test.go:639]: What is this required?

File: main_test.go [hunk 13]
```diff
@@ -544,6 +605,49 @@ func TestGetVMName(t *testing.T) {
...
+			http.DefaultServeMux = new(http.ServeMux)
+			http.HandleFunc("http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s", test.httpHandler)
+			want := true
```
Comment (easwars) [inline main_test.go:642]: Shouldn't you be using `test.wantOutput` instead?

File: main_test.go [hunk 14]
```diff
@@ -544,6 +605,49 @@ func TestGetVMName(t *testing.T) {
...
 }
 
+func TestCheckIPv6Capable(t *testing.T) {
+	tests := []struct {
+		desc        string
+		httpHandler func(http.ResponseWriter, *http.Request)
+		wantOutput  bool
+	}{
+		{
+			desc: "case with check on ipv6 enabled host",
```
Comment (easwars) [inline main_test.go:615]: See: go/go-style/decisions#subtest-names

I'm thinking `v6 enabled` and `v6 not enabled`.

File: main_test.go (hunk 13 above)
Comment (arvindbr8) [reply main_test.go]: yup thanks for catching that. when i fixed that it seems like there was a bug.. 
I was checking for err returned from `getFromMetadata`, but `getFromMetadata` does not return err for non-OK responses, and we return only the body from `getFromMetadata` -- so there is no way to check for status code from the caller. 
i'm adding addition logic in `getFromMetadata` to check for non-OK status codes. Lemeno what you think..

**Review by arvindbr8 — COMMENTED** [review]
rev'ing the PR addressing comments. @easwars pls take a look when you can

File: main.go [hunk 15]
```diff
@@ -347,6 +351,13 @@ func getVMName() string {
 	return string(vm)
 }
 
+// isIPv6Capable. Retrieve IPv6 address (if any) from metadata server to check
+// for DirectPath IPv6 capability.
```
Comment (easwars) [inline main.go:355]: How about:

```golang
// isIPv6Capable returns true if the VM is configured with an IPv6 address. 
// It contacts the metadata server to get this information.
```

File: main.go [hunk 16]
```diff
@@ -363,20 +374,23 @@ func getFromMetadata(urlStr string) ([]byte, error) {
 		},
 	}
 	resp, err := client.Do(req)
+	defer resp.Body.Close()
```
Comment (easwars) [inline main.go:377]: Why is this change required?

According to the `http` package documentation for `Do()`:
```
On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when
CheckRedirect fails, and even then the returned Response.Body is already closed.
```

File: main.go [hunk 17]
```diff
@@ -363,20 +374,23 @@ func getFromMetadata(urlStr string) ([]byte, error) {
...
 		return nil, fmt.Errorf("failed communicating with metadata server: %w", err)
 	}
 	body, err := ioutil.ReadAll(resp.Body)
-	resp.Body.Close()
 	if err != nil {
 		return nil, fmt.Errorf("failed reading from metadata server: %w", err)
 	}
+	if statusOk := resp.StatusCode >= 200 && resp.StatusCode < 300; !statusOk {
+		return nil, fmt.Errorf("failed reading from metadata server with Non-OK status code: %d", resp.StatusCode)
+	}
```
Comment (easwars) [inline main.go:387]: Maybe something like:

```golang
if code := resp.StatusCode; code < 200 || code > 299 {
	return nil, fmt.Errorf("metadata server returned status code %d for url %q", code, url)
}
```

I'm OK with the conditional being `!(code >= 200 && code < 300)` as well. But I feel the above one is simpler. Up to you.

File: main.go [hunk 18]
```diff
@@ -388,6 +402,35 @@ type server struct {
 	ServerFeatures []string `json:"server_features,omitempty"`
 }
 
+func generateServerConfigsFromInputs(serverUris []string, in configInput) []server {
```
Comment (easwars) [inline main.go:405]: I understand this is future proof in the sense that it will handle the case where we want to support multiple server configs. But we dont have that need today and we don't foresee it in the near future. So, I would recommend changing this function to be `func generateServerConfigsFromInputs(serverUri string, in configInput) []server { ...}`.

See: https://engdoc.corp.google.com/eng/doc/tott/episodes/438.md?cl=head

File: main_test.go [hunk 19]
```diff
@@ -544,6 +605,48 @@ func TestGetVMName(t *testing.T) {
...
+			},
+			wantOutput: false,
+		},
+	}
+	server := httptest.NewServer(nil)
+	defer server.Close()
+	overrideHTTP(server)
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			http.DefaultServeMux = new(http.ServeMux)
```
Comment (easwars) [inline main_test.go:639]: Do we still need this? It doesn't feel right to me to overwrite the package level variable `DefaultServeMux`.

If we really want to do something like this, it might involve a few more steps I think. The `httptest` package does not support using a serveMux other than the default one. So, we might have to create our own `net.Listener`, and pass it to [Serve](https://pkg.go.dev/net/http#Serve), along with a `Handler`. Would be nice if we can avoid all of that.

File: main.go (hunk 18 above)
Comment (arvindbr8) [reply main.go:405]: :+1: makes sense.. Im updating this as well

File: main_test.go (hunk 19 above)
Comment (arvindbr8) [reply main_test.go:639]: I had to do that because it throws an error since im trying to register a different handlerfunc for the same endpoint..

but this works

```
mux := new(http.ServeMux)
mux.HandleFunc("metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s", test.httpHandler)
server := httptest.NewServer(mux)
defer server.Close()
overrideHTTP(server)
if got := isIPv6Capable(); got != test.wantOutput {
 t.Fatalf("isIPv6Capable() = %t, want: %t", got, test.wantOutput)
}
```

lemeno if this is okayy

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
@ejona86 : I think the changes look good. I'm assigning you for a second set of eyes. Thank you.

File: main.go [hunk 20]
```diff
@@ -371,12 +382,15 @@ func getFromMetadata(urlStr string) ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed reading from metadata server: %w", err)
 	}
+	if code := resp.StatusCode; code < 200 || code > 299 {
+		return nil, fmt.Errorf("failed reading from metadata server with Non-OK status code: %d", code)
```
[the remark this answers is not in the document]
Comment (arvindbr8) [reply main.go:386]: yup that makes sense.. I'll update that

File: main.go (hunk 4 above)
Comment (ejona86) [reply main.go]: meh. Users should never need to set this. It'll just become the default and then go away.

**Review by ejona86 — APPROVED** [review]
